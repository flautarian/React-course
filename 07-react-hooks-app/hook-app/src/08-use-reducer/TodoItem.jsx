import React from 'react'

export const TodoItem = ({item, toDelete, toToggle}) => {

    const onRemoveTodo = (target) => {
        toDelete(item.id);
    }

    const onToogleDone = (target) => {
        toToggle(item.id);
    }

    return (
        <li className='list-group-item d-flex justify-content-between'>
            <span className='align-self-center'>{item.desc}</span>
            <div>
                <button className='btn btn-danger' onClick={ onRemoveTodo }>Delete</button>
                { item.done && <button className='btn btn-success' onClick={ onToogleDone }>Done</button>}
                { !item.done && <button className='btn btn-warning' onClick={ onToogleDone }>To do</button>}
            </div>
        </li>
    )
}
