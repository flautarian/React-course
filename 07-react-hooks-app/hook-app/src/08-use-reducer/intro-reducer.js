const initialState = [{
    id: 1,
    todo: 'Collect soul stone',
    done: false,
}];


const todoReducer = (state = initialState, action) => {
    if(!!action && action.type === "[TODO] add todo"){
        return [...state, action.payload];
    }
    return state;
}

const newTodo = {
    id: 2,
    todo: 'Collect power stone',
    done: false,
}

const addTodoAction = {
    type: "[TODO] add todo",
    payload: newTodo,
}

let todos = todoReducer();

todos = todoReducer(todos, addTodoAction);

console.log(todos);