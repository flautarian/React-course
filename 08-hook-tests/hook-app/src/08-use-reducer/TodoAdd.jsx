import React, { Fragment, useEffect, useState } from 'react'

export const TodoAdd = ({ createTodo }) => {

    const [form, setForm] = useState({
        desc: "",
        done: false
    });

    const { desc, done } = form;

    const onFormChange = ({ target }) => {
        const { name, value, checked, type } = target;
        setForm({
            ...form,
            [name]: type === "checkbox" ? checked : value
        })
    }

    useEffect(() => { }, [form]);

    const onNewTodo = (event) => {
        event.preventDefault();
        const todo = {
            id: new Date().getTime(),
            desc,
            done,
        }
        createTodo(todo);
    }

    return (
        <Fragment>
            <form onSubmit={(event) => onNewTodo(event)}>
                <input
                    type="text"
                    placeholder='What to do?'
                    className='form-control'
                    value={desc}
                    name='desc'
                    onChange={onFormChange} ></input>
                Done: 
                <input className='mt-2' type="checkbox" name='done' value={done} onChange={onFormChange} ></input>
                <br/>
                <button type='submit' className='btn btn-outline-primary mt-2'>Add Todo</button>
            </form>
        </Fragment>
    )
}
