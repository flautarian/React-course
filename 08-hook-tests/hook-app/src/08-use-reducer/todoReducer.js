

export const todoReducer = (initialState, action) => {
    switch (action.type) {
        case 'add':
            return [...initialState, action.payload];
        case 'remove':
            return initialState.filter(element => element.id !== action.id);
        case 'toggleDone':
            return initialState.map(element => {
                if (element.id === action.id) element.done = !element.done;
                return element;
            })
        default:
            return initialState;
    }
}