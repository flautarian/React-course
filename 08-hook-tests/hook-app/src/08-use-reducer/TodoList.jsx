import React, { useEffect } from 'react'
import { TodoItem } from './TodoItem'

export const TodoList = ({ list, toDelete, toToggle }) => {

    useEffect(() => { }, [list]);

    return (
        <ul className='list-group'>
            {
                list.map(todo => (
                    <TodoItem key={todo.id} item={todo} toDelete={toDelete} toToggle={toToggle} />
                ))
            }
        </ul>
    )
}
