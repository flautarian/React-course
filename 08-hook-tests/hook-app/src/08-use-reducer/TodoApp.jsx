import React, { Fragment, useEffect, useReducer } from 'react'
import { useTodos } from '../hooks/useTodos';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer'

export const TodoApp = () => {

    const { state, todosCount, todosRemaining, deleteTodo, toggleTodo, createTodo } = useTodos([]);

    return (
        <Fragment>
            <h2>TodoApp ({todosCount})
                <small>
                    Pendent: ({todosRemaining})
                </small>
            </h2>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodoList list={state} toDelete={deleteTodo} toToggle={toggleTodo} />
                </div>

                <div className="col-5">
                    <h4>Add todo Action:</h4>
                    <hr />
                    <TodoAdd createTodo={createTodo} />
                </div>
            </div>

        </Fragment>
    )
}
