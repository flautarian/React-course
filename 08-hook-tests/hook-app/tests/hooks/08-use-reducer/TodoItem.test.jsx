import { fireEvent, render, screen } from "@testing-library/react";
import { TodoItem } from "../../../src/08-use-reducer/TodoItem";

describe('TodoItem test', () => {
    const item = {
        id: 1,
        desc: "Todo Description",
        done: false
    };

    const mockToDelete = jest.fn();
    const mockToToggle = jest.fn();

    beforeEach(() => { jest.clearAllMocks(); });

    test('should match snapshot', () => {
        const { container } = render(<TodoItem item={item} />);
        expect(container).toMatchSnapshot();
    });

    test('should toggle the Todo item', () => {
        render(<TodoItem item={item} toToggle={mockToToggle} toDelete={mockToDelete} />);
        const toggleButton = screen.getByText('To do') || screen.getByText('Done');
        fireEvent.click(toggleButton);
        expect(mockToToggle).toHaveBeenCalledWith(item.id);
    });

    test('should delete the Todo item', () => {
        render(<TodoItem item={item} toToggle={mockToToggle} toDelete={mockToDelete} />);
        const deleteButton = screen.getByText('Delete');
        fireEvent.click(deleteButton);
        expect(mockToDelete).toHaveBeenCalledWith(item.id);
    });
});