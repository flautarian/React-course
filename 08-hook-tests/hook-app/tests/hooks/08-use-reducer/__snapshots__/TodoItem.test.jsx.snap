// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem test should match snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      class="align-self-center"
    >
      Todo Description
    </span>
    <div>
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
      <button
        class="btn btn-warning"
      >
        To do
      </button>
    </div>
  </li>
</div>
`;
